### DISCUSSION TITLE
drawBitmap() not always working...

### DISCUSSION THREAD

**Discussion by exdiogene** [fdivitto/FabGL#230] [General]
I was trying to enhance the solution for image drawing proposed in this discussion : 

https://github.com/fdivitto/FabGL/discussions/160

The solution provided by "e-labInnovations" is a very slow process taking more than 900mS to display a 256 X 256 picture.

I had better performance using this code instead of his : 
```
      r = ((my_data >> 11) & 0x1F) << 3;   // red   0 .. 255
      g = ((my_data >> 5) & 0x3F) << 2;  // green 0 .. 255
      b = (my_data & 0x1F) << 3;         // blue  0 .. 255

```
But it still take 630mS to display the picture!

So I tried to convert a line_array to a Bitmap_array to accelerate the process : 
```
    png.getLineAsRGB565(pDraw, bpm_data2, PNG_RGB565_LITTLE_ENDIAN, 0xffffffff);
    for (int16_t i = 0; i < pDraw->iWidth; i++) {
      my_data = bpm_data2[i];
      r = ((my_data >> 11) & 0x1F) << 3;   // red   0 .. 255
      g = ((my_data >> 5) & 0x3F) << 2;  // green 0 .. 255
      b = (my_data & 0x1F) << 3;         // blue  0 .. 255
      myBitmap_data[(i * 4)] = r;
      myBitmap_data[(i * 4) + 1] = g;
      myBitmap_data[(i * 4) + 2] = b;
      myBitmap_data[(i * 4) + 3] = 0xff;
    }
    Bitmap myBitmap = Bitmap(pDraw->iWidth, 1, &myBitmap_data[0], PixelFormat::RGBA8888);
    canvas.drawBitmap(0,pDraw->y,&myBitmap);
```

Now the process take only  65mS to display the picture, but the picture is partially displayed between 20% to 50% vertically...

Some horizontal lines get lost in the process, probably due by the VGA synchronization process, is there a way to display the missing lines correctly?

The only found solution at the moment is to put a "canvas.waitCompletion(false);" after the drawBitmap(), but it increase the process to 120mS.

I tried to put the "canvas.waitCompletion(false);" before the drawBitmap(), but it create some garbage in my picture, do you know why?

Any answer would be greatly appreciated...

**Comment by exdiogene** [comment]
I was able to reduce the time to 88mS by converting the Bitmap_array to the PixelFormat RGBA2222.

Now I wonder about the way to construct a Color Palette while proceeding the bytes in the array, my images have less than 16 colors. Is it possible to have a hint how to create and use a Palette with "drawBitmap()"?

**Comment by FreeLANMan** [comment]
If you are so concerned about speed it is better to use 'fabgl::VGA2Controller' and low resolutions.
For my 'PCEmulator Plus' I decided for the time being to use the tjpg lib, 400x300 and 64 colors at the cost of slowness (11 seconds per image). My problem is that some colors appear different in the final result and maybe it would be better if I knew the 'Color Profile' that would be ideal to convert the images.

**Comment by exdiogene** [comment]
I do not understand why you use the tjpg lib with a 11 seconds per image speed for 400x300?

While I was able to achieve 88mS with a PNG library and 256X256 images! (That would be 40 times faster with your images)

By the way I was using the 64 colors in my projects...
